### PR TITLE
Meta-set nodedef: Part 1

### PR DESCRIPTION
**Pull request by Thomas--S**
Mainly a refactor of parts of #1118.

Please can anyone tell me, if I'm on the correct track with this.

@Ekdohibs Could you please review this briefly, as it's mainly your code (adapted to current Minetest).

---
Edit: 

NOTE:

This PR does actually nothing on its own, it is only a first step for meta-set nodedef.

### CODE REVIEW

File: src/util/pointer.h [hunk 1]
```diff
@@ -180,7 +274,6 @@ class Buffer
 
 /************************************************
  *           !!!  W A R N I N G  !!!            *
- *           !!!  A C H T U N G  !!!            *
```
Comment (bigfoot547) [inline src/util/pointer.h]: Lol, why was this even here?

Comment (Thomas--S) [inline src/util/pointer.h]: 8560ece02e36b1e0ee7b86db2a38b8becbb639e4

**Comment by Thomas--S** [comment]
Can someone please tell me, why LINT fails?

**Comment by ShadowNinja** [comment]
@Thomas--S: If you look at the build logs you'll see why it's failing and exactly what has to change to fix it: http://sprunge.us/RCUQ?diff

**Comment by bigfoot547** [comment]
In order to fix node_with_def.h:
```diff

--- src/node_with_def.h	2017-06-03 15:35:53.399302389 +0000
+++ /dev/fd/63	2017-06-03 15:36:12.547303235 +0000
@@ -36,18 +36,9 @@
 	NodeWithDef(const MapNode &n, const INodeDefManager *ndef);
 	NodeWithDef(const MapNode &n, const HybridPtr<const ContentFeatures> &def);
 
-	const MapNode & node() const
-	{
-		return *this;
-	}
-	const ContentFeatures * def() const
-	{
-		return m_def.get();
-	}
-	bool def_is_global() const
-	{
-		return m_def.getNeverDelete();
-	}
+	const MapNode &node() const { return *this; }
+	const ContentFeatures *def() const { return m_def.get(); }
+	bool def_is_global() const { return m_def.getNeverDelete(); }
 
 	void setNode(const MapNode &n)
 	{
@@ -78,6 +69,4 @@
 	void getSelectionBoxes(std::vector<aabb3f> *boxes, u8 neighbors);
 };
 
-
-
 #endif
```

And then to fix node_with_def.cpp:
```diff
--- src/node_with_def.cpp	2017-06-03 15:35:53.399302389 +0000
+++ /dev/fd/63	2017-06-03 15:36:15.123303343 +0000
@@ -24,16 +24,14 @@
 NodeWithDef::NodeWithDef()
 {
 }
-NodeWithDef::NodeWithDef(const MapNode &n, const INodeDefManager *ndef):
-	MapNode(n),
-	m_def(NULL)
+NodeWithDef::NodeWithDef(const MapNode &n, const INodeDefManager *ndef)
+    : MapNode(n), m_def(NULL)
 {
 	m_def = &ndef->get(n);
 	m_def.setNeverDelete(true);
 }
-NodeWithDef::NodeWithDef(const MapNode &n, const HybridPtr<const ContentFeatures> &def):
-	MapNode(n),
-	m_def(def)
+NodeWithDef::NodeWithDef(const MapNode &n, const HybridPtr<const ContentFeatures> &def)
+    : MapNode(n), m_def(def)
 {
 }
 
@@ -96,4 +94,3 @@
 	const ContentFeatures &f = *m_def;
 	return MapNode::getSelectionBoxes(f, boxes, neighbors);
 }
-
```
@Thomas--S

**Comment by bigfoot547** [comment]
In order to find diffs like the above ones, check the [failed travis build](https://travis-ci.org/minetest/minetest/jobs/239098719).

**Comment by Thomas--S** [comment]
@bigfoot547 I already found that diffs, but I didn't believe that this the diff I have to use for our codestyle. I didn't believe that we collapse these things in node_with_def.h to one line.

**Comment by bigfoot547** [comment]
I didn't know I needed to squish [these comma things](https://github.com/minetest/minetest/blob/master/src/script/lua_api/l_camera.cpp#L194-L196).

**Comment by Ekdohibs** [comment]
@Thomas--S I looked at the code, and it seems ok and consistent with what I remember. Thanks for rebasing the PR! Last time I tried to rebase it, I tried to extract the NodeWithDef changes, but since I tried to extract more of them, I didn't complete the rebase.

I think the next step after that will be to try to use NodeWithDef in at least the following places:
- light spreading
- physics & collision detection
- maybe pathfinding? that was not done in my code though
However, you should probably *not* use it in rendering, at the function is performance critical and some special handling would make it more efficient, I think. I sincerely hope that once you get to the meta_set_nodedef part, you manage to avoid the performance problems I had (which were probably caused by a copy I didn't notice somewhere, I am not sure).

**Comment by Thomas--S** [comment]
Thank you very much for your explanation, Ekdohibs.

Maybe this PR could be merged soon, although it is quite useless as stand-alone, but this will surely help us to avoid endless rebases.

**Comment by Thomas--S** [comment]
I think the WIP tag can be removed here because part 1 seems to be quite ready.

**Comment by Thomas--S** [comment]
PR is rebased.

I think one is ready to be reviewed (and merged).

File: src/util/pointer.h [hunk 2]
```diff
@@ -25,6 +25,96 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include <cstring>
 
 template <typename T>
+class HybridPtr
```
Comment (nerzhul) [inline src/util/pointer.h]: What is the purpose of this classe ? unique_ptr + weak_ptr ?

**Comment by Thomas--S** [comment]
Please don't merge this yet, there's still a bug with rotation. I'll try to fix it tomorrow.

**Comment by Thomas--S** [comment]
The problems are fixed now, they were located in part 3. Part 1 - 3 are now ready for review again.

File: src/mapnode.cpp [hunk 3]
```diff
@@ -182,8 +207,7 @@ v3s16 MapNode::getWallMountedDir(INodeDefManager *nodemgr) const
 	}
 }
 
-void MapNode::rotateAlongYAxis(INodeDefManager *nodemgr, Rotation rot)
-{
+void MapNode::rotateAlongYAxis(INodeDefManager *nodemgr, Rotation rot) {
```
Comment (Ekdohibs) [inline src/mapnode.cpp]: Shouldn't that brace stay on the next line?

File: src/mapnode.h [hunk 4]
```diff
@@ -230,10 +239,20 @@ struct MapNode
...
 	// 0 <= return value <= LIGHT_SUN
+	u8 getLightBlend(u32 daylight_factor, const ContentFeatures &f) const
+	{
+		u8 lightday = 0;
+		u8 lightnight = 0;
+		getLightBanks(lightday, lightnight, f);
+		return blend_light(daylight_factor, lightday, lightnight);
+	}
+
 	u8 getLightBlend(u32 daylight_factor, INodeDefManager *nodemgr) const
```
Comment (Ekdohibs) [inline src/mapnode.h]: Is there any reason this function isn't factorized with the previous one?

**Pull request closed without merging**